src/coreclr/vm/eetoprofinterfaceimpl.cpp

```diff
@@ -102,8 +102,32 @@ enum ClrToProfEntrypointFlags
     kEE2PNoTrigger                      = 0x00000004,
 };
 
-#define ASSERT_EVAC_COUNTER_NONZERO()   \
-    _ASSERTE(m_pProfilerInfo->dwProfilerEvacuationCounter.Load() > 0)
+EvacuationCounterHolder::EvacuationCounterHolder(ProfilerInfo *pProfilerInfo) :
+    m_pProfilerInfo(pProfilerInfo),
+    m_pThread(GetThreadNULLOk())
+{
+    _ASSERTE(m_pProfilerInfo != NULL);
+    if (m_pThread == NULL)
```

Comment:
Does this mean that the profiler detach can crash if it happens to interleave with a callback issued where GetThread is null?

How many of the callbacks are affected? Did we have this problem before?

Reply:
For all of desktop and before my change in coreclr the evacuation counters were on the Thread * and it would only increment/decrement if GetThread returned non-null. So this behavior should be identical to 5.0 and previous runtimes.

I have been thinking since last night about if there was a latent bug. Mostly I don't think so, but there is a couple places like EventPipe callbacks that may not have a managed thread

Reply:
yeah, I am pretty suspicious there is a latent bug here when it comes to EventPipe notifications - they can be triggered on threads where GetThread() == NULL

Reply:
@noahfalk thinking about it though, with the profiler they are synchronous on the thread that it happens. We may not actually have a bug, since I would expect any place EventPipe events happen to have an EE thread. Since we have the managed bridge from native events to EventPipe.

Reply:
This issue also affects server GC threads. Server GC threads do not have `Thread*`.

Reply:
> I would expect any place EventPipe events happen to have an EE thread

I'm ~98% sure this isn't the case : ) Some of the EventPipe events are emitted where GetThread()==NULL (such as the events from Server GC threads @jkotas mentioned). Originally the code in EventPipe's WriteEvent() code refered to GetThread() for the thread-local write buffers but that was broken. We had to invent the new EventPipeThread object to work around the lack of `Thread*` for some events.

Reply:
Ok, I agree there is a latent issue here, but it also has been present ever since the EventPipe stuff was merged in 3.1. Given that and that there's also an easy workaround - stop the session before detaching -  I propose I file a bug and don't try to solve this now.

https://github.com/dotnet/runtime/issues/60081

Reply:
Yeah, I wasn't pushing to have you fix it now given that it was pre-existing, just confirming that it does exist.